File: Leetcode/linkedlist/328_OddEvenLinkedList/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode oddEvenList(ListNode head) {
        /*///original solution
        if(head==null) return head;
        ListNode oddHead = new ListNode(-1), temp = head, tail = temp, odd = oddHead;
        while(temp!=null){
            odd.next = temp.next;
            odd = odd.next;
            if(temp.next!=null){
                temp.next = temp.next.next;
                temp = temp.next;
            }else{
                temp.next = null;
            }
            temp = temp.next;
        }
        temp = head;
        while( temp.next!=null){ temp = temp.next; }
        temp.next = oddHead.next;
        return head;/*/
        if(head==null) return null;
        ListNode odd = head, evenHead = head.next, even = evenHead;
        while(even!=null && even.next!=null){
            odd.next = odd.next.next;
            even.next = even.next.next;
            odd = odd.next;
            even = even.next;
        }
        odd.next = evenHead;
        return head;
    }
}

File: Leetcode/linkedlist/328_OddEvenLinkedList/Solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        if(head!=nullptr){
            ListNode* odd = head;
            ListNode* evenHead = head->next;
            ListNode* even = evenHead;
            while(even!=nullptr && even->next!=nullptr){
                odd->next = odd->next->next;
                even->next = even->next->next;
                odd = odd->next;
                even = even->next;
            }
            odd->next = evenHead;
        }
        return head;
    }
};
